src/lib.rs

```diff
@@ -2,106 +2,167 @@
 #![cfg_attr(test, deny(warnings))]
 #![cfg_attr(feature = "heap_size", feature(custom_derive, plugin))]
 #![cfg_attr(feature = "heap_size", plugin(heapsize_plugin))]
+#![cfg_attr(feature = "nightly", feature(test))]
 
-//! # Case
+//! # UniCase
 //!
-//! Case provices a way of specifying strings that are case-insensitive.
+//! UniCase provices a way of specifying strings that are case-insensitive.
+//!
+//! UniCase supports full [Unicode case
+//! folding](https://www.w3.org/International/wiki/Case_folding). It can also
+//! utilize faster ASCII case comparisons, if both strings are ASCII.
+//!
+//! Using the `UniCase::new()` constructor will check the string to see if it
+//! is all ASCII. When a `UniCase` is compared against another, if both are
+//! ASCII, it will use the faster comparison.
+//!
+//! There also exists the `Ascii` type in this crate, which will always assume
+//! to use the ASCII case comparisons, if the encoding is already known.
 //!
 //! ## Example
 //!
 //! ```rust
 //! use unicase::UniCase;
 //!
-//! let a = UniCase("foobar");
-//! let b = UniCase("FoObAr");
+//! let a = UniCase::new("Maße");
+//! let b = UniCase::new("MASSE");
+//! let c = UniCase::new("mase");
+//!
+//! assert_eq!(a, b);
+//! assert!(b != c);
+//! ```
+//!
+//! ## Ascii
+//!
+//! ```rust
+//! use unicase::Ascii;
+//!
+//! let a = Ascii::new("foobar");
+//! let b = Ascii::new("FoObAr");
 //!
 //! assert_eq!(a, b);
 //! ```
 
 #[cfg(feature = "heap_size")]
 extern crate heapsize;
+#[cfg(feature = "nightly")]
+extern crate test;
 
-use std::ascii::AsciiExt;
 #[cfg(iter_cmp)]
 use std::cmp::Ordering;
 use std::fmt;
 use std::hash::{Hash, Hasher};
 use std::ops::{Deref, DerefMut};
 use std::str::FromStr;
 
+use self::unicode::Unicode;
+
+mod ascii;
+mod unicode;
+
 /// Case Insensitive wrapper of strings.
-#[derive(Copy, Clone, Debug)]
+#[derive(Clone, Copy)]
+#[cfg_attr(feature = "heap_size", derive(HeapSizeOf))]
+pub struct UniCase<S>(Encoding<S>);
+
+/// Case Insensitive wrapper of Ascii strings.
+#[derive(Clone, Copy, Debug)]
+#[cfg_attr(feature = "heap_size", derive(HeapSizeOf))]
+pub struct Ascii<S>(S);
+
+
+#[derive(Clone, Copy, Debug)]
 #[cfg_attr(feature = "heap_size", derive(HeapSizeOf))]
-pub struct UniCase<S>(pub S);
+enum Encoding<S> {
+    Ascii(Ascii<S>),
+    Unicode(Unicode<S>),
+}
+
+macro_rules! inner {
+
+    (mut $e:expr) => ({
+        match &mut $e {
+            &mut Encoding::Ascii(ref mut s) => &mut s.0,
+            &mut Encoding::Unicode(ref mut s) => &mut s.0,
+        }
+    });
+    ($e:expr) => ({
+        match &$e {
+            &Encoding::Ascii(ref s) => &s.0,
+            &Encoding::Unicode(ref s) => &s.0,
+        }
+    });
+}
+
+impl<S: AsRef<str>> UniCase<S> {
+    pub fn new(s: S) -> UniCase<S> {
```

Comment:
Would be good to add the module doc that this constructor involves a scan to here.